lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -1402,9 +3612,54 @@ class Differentiation : public SILModuleTransform {
 public:
   Differentiation() : SILModuleTransform() {}
   void run() override;
+private:
+  void processGradientInst(GradientInst *gi, ADContext &context);
 };
 } // end anonymous namespace
 
+void Differentiation::processGradientInst(GradientInst *gi,
+                                          ADContext &context) {
+  SILFunction *parent = gi->getFunction();
+  auto operand = gi->getOperand(0);
+  SILFunction *gradFn = nullptr;
+  // If it traces back to a `function_ref`, differentiate that.
+  if (auto *originalFRI = findReferenceToVisibleFunction(operand)) {
+    auto *original = originalFRI->getReferencedFunction();
+    gradFn = lookupOrSynthesizeGradient(context, gi, original);
+    
+    // If `gradFn` is still null, errors must have occurred.
+    if (!gradFn) return;
+    
+    // Replace the `gradient` instruction with the reference to the specified
+    // function.
+    SILBuilder builder(gi);
+    auto loc = parent->getLocation();
+    SILValue gradRef = builder.createFunctionRef(loc, gradFn);
+    // Traverse from the `gradient` instruction to the original
+    // `function_ref`. If there's any function convertion, do the same
+    // convertion for the gradient.
+    auto convertedGradFn = reapplyFunctionConversion(gradRef, originalFRI,
+                                                     gi->getOriginal(),
+                                                     builder, loc);
+    // Replace uses of the `gradient` instruction with the converted (if
+    // necessary) gradient function value.
+    gi->replaceAllUsesWith(convertedGradFn);
+  }
+  // Differentiating opaque functions is not supported yet.
+  else {
+    if (auto loc = gi->getLoc()) {
+      auto *expr = loc.castToASTNode<ReverseAutoDiffExpr>();
+      context.diagnose(expr->getOriginalExpr()->getLoc(),
+                       diag::autodiff_opaque_function_unsupported);
+    }
+    context.setErrorOccurred();
+    return;
+  }
+  // We invalide analyses on the parent function because the `gradient`
```

Comment:
Typos: `invalide` -> `invalidate`, on the next line `instruciton` -> `instruction`.

Reply:
Done.